File: network-library/src/main/java/cn/maoyanluo/network_library/tcp/TcpSocketServer.kt
package cn.maoyanluo.network_library.tcp

import cn.maoyanluo.coroutine_library.CoroutineManager
import cn.maoyanluo.socket_common_library.SocketServer
import cn.maoyanluo.socket_common_library.SocketServerCallback
import java.io.InputStream
import java.io.OutputStream
import java.net.ServerSocket
import java.net.Socket

class TcpSocketServer(
    private val port: Int,
    serverCallback: SocketServerCallback<Socket>,
    coroutineManager: CoroutineManager
): SocketServer<ServerSocket, Socket>(serverCallback, coroutineManager) {

    override fun createServerSocket(): ServerSocket = ServerSocket(port)

    override fun acceptSocket(serverSocketSnapshot: ServerSocket?): Socket? = serverSocketSnapshot?.accept()

    override fun createAcceptClient(
        socket: Socket,
        callback: SocketServerCallback.ClientCallback,
        coroutineManager: CoroutineManager
    ): SocketServer.Client<Socket> = Client(
        socket,
        callback,
        coroutineManager
    )

    class Client(
        private val socket: Socket,
        callback: SocketServerCallback.ClientCallback,
        coroutineManager: CoroutineManager
    ) : SocketServer.Client<Socket>(socket, callback, coroutineManager) {

        override fun getOutputStream(): OutputStream? = socket.getOutputStream()
        override fun getInputStream(): InputStream? = socket.getInputStream()

    }

}

File: Windows/GameControllerSimulator/NetworkLibrary/TcpSocketServer.cs
using SocketCommonLibrary;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Threading.Tasks;

namespace NetworkLibrary
{
    public class TcpSocketServer : SocketServer<TcpListener, TcpClient>
    {
        public override ConnectionType ConnectionType => ConnectionType.TCP;
        private int port;
        public TcpSocketServer(int port, SocketServerCallback<TcpListener, TcpClient> callback) : base(callback)
        {
            this.port = port;
        }

        protected override async Task<TcpClient> AcceptSocket(TcpListener serverSocker)
        {
            return await serverSocker.AcceptTcpClientAsync();
        }

        protected override void Close(TcpListener serverSocker)
        {
            serverSocker.Stop();
            serverSocker.Dispose();
        }

        protected override SocketClient<TcpClient> CreateAcceptClient(TcpClient socket, SocketClientCallback<TcpClient> clientCallback)
        {
            return new TcpSocketClient(socket, clientCallback);
        }

        protected override TcpListener CreateServerSocket()
        {
            TcpListener listener = new TcpListener(IPAddress.Any, port);
            listener.Start();
            return listener;
        }
    }

    public class TcpSocketClient : SocketClient<TcpClient>
    {
        public override ConnectionType ConnectionType => ConnectionType.TCP;
        public TcpSocketClient(TcpClient socket, SocketClientCallback<TcpClient> callback) : base(socket, callback)
        {
        }

        protected override void Close(TcpClient socket)
        {
            socket.Close();
            socket.Dispose();
        }

        protected override Stream GetStream(TcpClient socket)
        {
            return socket.GetStream();
        }
    }
}
